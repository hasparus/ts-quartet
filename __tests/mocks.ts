import {
  FunctionSchema,
  IVariantSchema,
  Schema,
  IObjectSchema
} from "../types";
import { v } from "../index";

export const primitives: any[] = [
  undefined,
  null,
  true,
  false,
  0,
  1,
  NaN,
  Infinity,
  -Infinity,
  "",
  "Andrew",
  Symbol.for("test")
];

export const funcSchema: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`
});
export const funcSchemaWithNot: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`,
  not: value => `(typeof ${value} !== 'number' || ${value} % 2 !== 0)`
});
export const funcSchemaWithHandleError: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`
});
export const funcSchemaWithNotHandleError: FunctionSchema = () => ({
  check: value => `typeof ${value} === 'number' && ${value} % 2 === 0`,
  not: value => `(typeof ${value} !== 'number' || ${value} % 2 !== 0)`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`
});
export const funcSchemaWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});
export const funcSchemaWithNotWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  not: (value, ctx) =>
    `(typeof ${value} !== 'number' || ${value} % ${ctx}.divide !== 0)`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});
export const funcSchemaWithHandleErrorWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});
export const funcSchemaWithNotHandleErrorWithPrepare: FunctionSchema = () => ({
  check: (value, ctx) =>
    `typeof ${value} === 'number' && ${value} % ${ctx}.divider === 0`,
  handleError: (value, ctx) => `${ctx}.explanations.push(${value})`,
  not: (value, ctx) =>
    `(typeof ${value} !== 'number' || ${value} % ${ctx}.divider !== 0)`,
  prepare: ctx => {
    ctx.divider = 2;
  }
});

export const funcSchemas: FunctionSchema[] = [
  funcSchema,
  funcSchemaWithNot,
  funcSchemaWithHandleError,
  funcSchemaWithNotHandleError,
  funcSchemaWithPrepare,
  funcSchemaWithNotWithPrepare,
  funcSchemaWithHandleErrorWithPrepare,
  funcSchemaWithNotHandleErrorWithPrepare
];

const has = (obj: any, key: string | number) => {
  return Object.prototype.hasOwnProperty.call(obj, key);
};

export function* objects(
  level: number = 0,
  maxLevel: number = 2
): Iterable<object | any[]> {
  if (level > maxLevel) return;
  yield {};
  yield [];
  for (const obj of objects(level + 1, maxLevel)) {
    if (has(obj, "push")) {
      for (const primitive of primitives) {
        yield [...(obj as any), primitive];
      }
      for (const o of objects(level + 1, maxLevel)) {
        yield [...(obj as any), o];
      }
    } else {
      for (const primitive of primitives) {
        if (!has(obj, "even")) {
          yield { ...obj, even: primitive };
        }
        if (!has(obj, "another")) {
          yield { ...obj, another: primitive };
        }
      }
      for (const o of objects(level + 1, maxLevel)) {
        if (!has(obj, "even")) {
          yield { ...obj, even: o };
        }
        if (!has(obj, "another")) {
          yield { ...obj, another: o };
        }
        if (!has(obj, "nested")) {
          yield { ...obj, nested: o };
        }
      }
    }
  }
}

export function* variants(
  level: number = 0,
  maxLevel: number = 2
): Iterable<IVariantSchema> {
  if (level > maxLevel) {
    return;
  }
  yield [];
  for (const variant of variants(level + 1)) {
    for (const primitive of primitives) {
      yield [...variant, primitive];
    }
    for (const funcSchema of funcSchemas) {
      yield [...variant, funcSchema];
    }
    for (const obj of objectSchemas(level + 1, maxLevel)) {
      yield [...variant, obj];
    }
    for (const variant2 of variants(level + 1, maxLevel)) {
      yield [...variant, variant2];
    }
  }
}

export function* objectSchemas(
  level: number = 0,
  maxLevel: number = 2
): Iterable<IObjectSchema> {
  if (level > maxLevel) return;
  yield {};
  for (const obj of objectSchemas(level + 1, maxLevel)) {
    if (!has(obj, "even")) {
      for (const primitive of primitives) {
        yield { ...obj, even: primitive };
      }
      for (const funcSchema of funcSchemas) {
        yield { ...obj, even: funcSchema };
      }
      for (const obj of objectSchemas(level + 1, maxLevel)) {
        yield { ...obj, even: obj };
      }
      for (const obj of variants(level + 1, maxLevel)) {
        yield { ...(obj as any), even: obj };
      }
    }
    if (!has(obj, "nested")) {
      for (const primitive of primitives) {
        yield { ...obj, nested: primitive };
      }
      for (const funcSchema of funcSchemas) {
        yield { ...obj, nested: funcSchema };
      }
      for (const obj of objectSchemas(level + 1, maxLevel)) {
        yield { ...obj, nested: obj };
      }
      for (const obj of variants(level + 1, maxLevel)) {
        yield { ...(obj as any), nested: obj };
      }
    }
    if (!has(obj, v.rest)) {
      for (const primitive of primitives) {
        yield { ...obj, [v.rest]: primitive };
      }
      for (const funcSchema of funcSchemas) {
        yield { ...obj, [v.rest]: funcSchema };
      }
      for (const obj of objectSchemas(level + 1, maxLevel)) {
        yield { ...obj, [v.rest]: obj };
      }
      for (const obj of variants(level + 1, maxLevel)) {
        yield { ...(obj as any), [v.rest]: obj };
      }
    }
  }
}

export function* schemas(maxLevel: number): Iterable<Schema> {
  yield* primitives[Symbol.iterator]();
  yield* funcSchemas[Symbol.iterator]();
  yield* objectSchemas(0, maxLevel);
  yield* variants(0, maxLevel);
}

export function* values(maxLevel: number) {
  yield* primitives[Symbol.iterator]();
  yield* objects(0, maxLevel);
}
