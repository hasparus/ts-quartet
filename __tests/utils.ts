import { getDescription } from "./getDescription";

export function snapshot(validator: any) {
  const description = getDescription(validator);

  expect(description).toMatchSnapshot();
}
export function getExplanation(
  validator: { (arg0: any): void; explanations: any },
  value: any
) {
  validator(value);
  return validator.explanations;
}
export function isValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (!valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(true);
}
export function isNotValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(false);
}

export function tables(validator: any, valids: any[], invalids: any[]) {
  for (const valid of valids) {
    isValid(validator, valid);
  }
  for (const invalid of invalids) {
    isNotValid(validator, invalid);
  }
}
