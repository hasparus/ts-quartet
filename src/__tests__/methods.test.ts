import { v } from "../index";
import { snapshot, puretables } from "./utils";
import { primitives } from "./mocks";

describe("methods", () => {
  test("and", () => {
    expect(typeof v.and).toBe("function");
    const validator = v(v.and(v.number, v.custom(e => e % 2 === 0)));
    snapshot(validator);
    puretables(
      validator,
      [2, 4, 6, 8],
      [
        "2",
        "4",
        "6",
        ...primitives.filter(e => typeof e !== "number"),
        1,
        3,
        5,
        7,
        9
      ]
    );
  });
  test("arrayOf", () => {
    expect(typeof v.arrayOf).toBe("function");
    const validator = v(v.arrayOf(v.number));
    snapshot(validator);
    puretables(validator, [[], [1], [1, 2, 3.14]], [["1"], ...primitives]);
  });
  test("boolean", () => {
    expect(typeof v.boolean).toBe("function");
    const validator = v(v.boolean);
    snapshot(validator);
    puretables(
      validator,
      [...primitives.filter(e => typeof e === "boolean")],
      [...primitives.filter(e => typeof e !== "boolean"), [], {}]
    );
  });
  test("compileAnd", () => {
    expect(typeof v.compileAnd).toBe("function");
  });
  test("compileArrayOf", () => {
    expect(typeof v.compileArrayOf).toBe("function");
  });
  test("custom", () => {
    expect(typeof v.custom).toBe("function");
  });
  test("function", () => {
    expect(typeof v.function).toBe("function");
    const validator = v(v.function);
    snapshot(validator);
    puretables(validator, [() => {}, function() {}], [...primitives, [], {}]);
  });
  test("max exclusive", () => {
    expect(typeof v.max).toBe("function");
    const validator = v(v.max(5, true));
    puretables(
      validator,
      [1, 2, 3, 4, -Infinity],
      [5, 6, 7, 8, 9, 5.1, Infinity]
    );
    snapshot(validator);
  });
  test("max inclusive", () => {
    const validator = v(v.max(5));
    puretables(
      validator,
      [1, 2, 3, 4, -Infinity, 5],
      [6, 7, 8, 9, 5.1, Infinity]
    );
    snapshot(validator);
  });
  test("maxLength exclusive", () => {
    expect(typeof v.maxLength).toBe("function");
    const validator = v(v.maxLength(5, true));
    puretables(
      validator,
      ["1234", ...[1, 2, 3, 4].map(length => Array.from({ length }, () => 1))],
      [
        "12345",
        ...[6, 7, 8, 9, 5].map(length => Array.from({ length }, () => 1)),
        "123456"
      ]
    );
    snapshot(validator);
  });
  test("maxLength inclusive", () => {
    const validator = v(v.maxLength(5));
    puretables(
      validator,
      [
        "1234",
        "12345",
        ...[1, 2, 3, 4, 5].map(length => Array.from({ length }, () => 1))
      ],
      [...[6, 7, 8, 9].map(length => Array.from({ length }, () => 1)), "123456"]
    );
    snapshot(validator);
  });
  test("min exclusive", () => {
    expect(typeof v.min).toBe("function");
    const validator = v(v.min(5, true));
    puretables(
      validator,
      [6, 7, 8, 9, 5.1, Infinity],
      [1, 2, 3, 4, 5, -Infinity]
    );
    snapshot(validator);
  });
  test("min inclusive", () => {
    expect(typeof v.min).toBe("function");
    const validator = v(v.min(5));
    puretables(
      validator,
      [5, 6, 7, 8, 9, 5.1, Infinity],
      [1, 2, 3, 4, -Infinity]
    );
    snapshot(validator);
  });
  test("minLength exclusive", () => {
    expect(typeof v.minLength).toBe("function");
    const validator = v(v.minLength(5, true));
    puretables(
      validator,
      [
        ...[6, 7, 8, 9].map(length => Array.from({ length }, () => 1)),
        "123456"
      ],
      [
        "12345",
        "1234",
        ...[1, 2, 3, 4, 5].map(length => Array.from({ length }, () => 1))
      ]
    );
    snapshot(validator);
  });
  test("minLength inclusive", () => {
    expect(typeof v.minLength).toBe("function");
    const validator = v(v.minLength(5));
    puretables(
      validator,
      [
        ...[6, 7, 8, 9].map(length => Array.from({ length }, () => 1)),
        "123456",
        "12345"
      ],
      ["1234", ...[1, 2, 3, 4].map(length => Array.from({ length }, () => 1))]
    );
    snapshot(validator);
  });
  test("negative", () => {
    expect(typeof v.negative).toBe("function");
    const validator = v(v.negative);
    snapshot(validator);
    puretables(validator, [-1, -2, -Infinity], [0, -0, 1, 2, 3, 4, Infinity]);
  });
  test("number", () => {
    expect(typeof v.number).toBe("function");
    const validator = v(v.number);
    snapshot(validator);
    puretables(
      validator,
      [-1, -2, -Infinity, NaN, 0, -0, 1, 2, 3, 4, Infinity],
      [...primitives.filter(e => typeof e !== "number"), {}, []]
    );
  });
  test("not", () => {
    expect(typeof v.not).toBe("function");
  });
  test("positive", () => {
    expect(typeof v.positive).toBe("function");
    const validator = v(v.positive);
    snapshot(validator);
    puretables(validator, [1, 2, Infinity], [0, -0, -1, -2, -3, -4, -Infinity]);
  });
  test("rest", () => {
    expect(v.rest).toBe("__quartet/rest__");
  });
  test("restOmit", () => {
    expect(v.restOmit).toBe("__quartet/rest-omit__");
  });
  test("safeInteger", () => {
    expect(typeof v.safeInteger).toBe("function");
    const validator = v(v.safeInteger);
    snapshot(validator);
    puretables(
      validator,
      [1, 2, 1e8, 0, -1, -2, -3],
      [Math.PI, 1e40, NaN, Infinity, -Infinity]
    );
  });
  test("string", () => {
    expect(typeof v.string).toBe("function");
    const validator = v(v.string);
    snapshot(validator);
    puretables(
      validator,
      [...primitives.filter(e => typeof e === "string")],
      [...primitives.filter(e => typeof e !== "string"), [], {}]
    );
  });
  test("symbol", () => {
    expect(typeof v.symbol).toBe("function");
    const validator = v(v.symbol);
    snapshot(validator);
    puretables(
      validator,
      [...primitives.filter(e => typeof e === "symbol")],
      [...primitives.filter(e => typeof e !== "symbol"), [], {}]
    );
  });
  test("test", () => {
    expect(typeof v.test).toBe("function");
    const validator = v(v.test(/[Aa]nd/));
    snapshot(validator);
    puretables(
      validator,
      ["Andrew", "and", "demand"],
      ["", "ndrew", "nd", "deman"]
    );
  });
});
