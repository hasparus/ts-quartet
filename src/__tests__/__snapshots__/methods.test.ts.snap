// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`methods and 1`] = `
Object {
  "_": "function validator(value) {
  return validator.and(value)
}",
  "and": "function validator(value) {
  if (typeof value !== 'number') return false
  if (!validator.custom(value)) return false
  return true
}",
  "and.custom": "function (e) { return e % 2 === 0; }",
  "and.explanations": Array [],
  "and.pure": true,
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods arrayOf 1`] = `
Object {
  "_": "function validator(value) {
  return validator.arr(value)
}",
  "arr": "function validator(value) {
  if (!value || !Array.isArray(value)) return false
  for (let i = 0; i < value.length; i++) {
    const elem = value[i]
    if (typeof elem !== 'number') return false
  }
  return true
}",
  "arr.explanations": Array [],
  "arr.pure": true,
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods boolean 1`] = `
Object {
  "_": "function validator(value) {
  return typeof value === 'boolean'
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods function 1`] = `
Object {
  "_": "function validator(value) {
  return typeof value === 'function'
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods max exclusive 1`] = `
Object {
  "_": "function validator(value) {
  return value < 5
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods max inclusive 1`] = `
Object {
  "_": "function validator(value) {
  return value <= 5
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods maxLength exclusive 1`] = `
Object {
  "_": "function validator(value) {
  return value && value.length < 5
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods maxLength inclusive 1`] = `
Object {
  "_": "function validator(value) {
  return value && value.length <= 5
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods min exclusive 1`] = `
Object {
  "_": "function validator(value) {
  return value > 5
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods min inclusive 1`] = `
Object {
  "_": "function validator(value) {
  return value >= 5
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods minLength exclusive 1`] = `
Object {
  "_": "function validator(value) {
  return value && value.length > 5
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods minLength inclusive 1`] = `
Object {
  "_": "function validator(value) {
  return value && value.length >= 5
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods negative 1`] = `
Object {
  "_": "function validator(value) {
  return value < 0
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods number 1`] = `
Object {
  "_": "function validator(value) {
  return typeof value === 'number'
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods positive 1`] = `
Object {
  "_": "function validator(value) {
  return value > 0
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods safeInteger 1`] = `
Object {
  "_": "function validator(value) {
  return Number.isSafeInteger(value)
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods string 1`] = `
Object {
  "_": "function validator(value) {
  return typeof value === 'string'
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods symbol 1`] = `
Object {
  "_": "function validator(value) {
  return typeof value === 'symbol'
}",
  "explanations": Array [],
  "pure": true,
}
`;

exports[`methods test 1`] = `
Object {
  "_": "function validator(value) {
  return validator.tester.test(value)
}",
  "explanations": Array [],
  "pure": true,
  "tester": /\\[Aa\\]nd/,
}
`;
